File: src/engine/Game.ts
import { Entity } from './Entity';
import type { EntityStats, Team } from './Entity';
import { Vector2 } from './Vector2';
import { CONFIG } from './config';
import { Cards } from './Cards';

export interface Projectile {
    id: number;
    sourceId: string;
    startPos: Vector2;
    pos: Vector2;
    target: Entity;
    damage: number;
    speed: number;
    team: 'blue' | 'red';
    onHit?: () => void;
}

export interface VisualEffect {
    id: number;
    name: string;
    pos: Vector2;
    startTime: number;
    duration: number; // in seconds
}

export class Game {
    entities: Entity[] = [];
    projectiles: Projectile[] = [];
    effects: VisualEffect[] = [];
    nextEntityId: number = 1;
    nextProjectileId: number = 1;
    nextEffectId: number = 1;
    timeElapsed: number = 0;

    constructor() {}

    start() {
        // Blue Team (Bottom)
        this.addEntity(Cards.king_tower, 'blue', new Vector2(9, 29)); // Center back
        this.addEntity(Cards.princess_tower, 'blue', new Vector2(3, 25)); // Left
        this.addEntity(Cards.princess_tower, 'blue', new Vector2(15, 25)); // Right

        // Red Team (Top)
        this.addEntity(Cards.king_tower, 'red', new Vector2(9, 3)); // Center back
        this.addEntity(Cards.princess_tower, 'red', new Vector2(3, 7)); // Left
        this.addEntity(Cards.princess_tower, 'red', new Vector2(15, 7)); // Right


    }

    addEntity(stats: EntityStats, team: Team, pos: Vector2): Entity {
        // If it's a swarm card, spawn multiple
        if (stats.spawnCount && stats.spawnCount > 1) {
            let firstEntity: Entity | null = null;
            const radius = stats.spawnRadius || 1.0;
            const swarmGroupId = Math.random().toString(36).substring(7); // Shared group ID for the swarm
            for (let i = 0; i < stats.spawnCount; i++) {
                // Simple circular spread
                const angle = (i / stats.spawnCount) * Math.PI * 2;
                const offset = new Vector2(Math.cos(angle) * radius, Math.sin(angle) * radius);
                const e = new Entity(this.nextEntityId++, stats, team, pos.add(offset), this);
                e.cloneGroupId = swarmGroupId; // Override the constructor's random ID
                this.entities.push(e);
                if (!firstEntity) firstEntity = e;
            }
            return firstEntity!; // Returning one representative entity
        } else {
            const entity = new Entity(this.nextEntityId++, stats, team, pos, this);
            this.entities.push(entity);
            return entity;
        }
    }

    addEntityById(id: string, team: Team, pos: Vector2): Entity | null {
        const stats = Cards[id as keyof typeof Cards];
        if (!stats) return null;
        return this.addEntity(stats, team, pos);
    }

    update(dt: number) {
        this.timeElapsed += dt;

        // Update entities
        for (const entity of this.entities) {
            entity.update(dt);
        }

        // Update projectiles
        for (let i = this.projectiles.length - 1; i >= 0; i--) {
            const p = this.projectiles[i];
            
            // If target is dead, projectile still travels to last known pos, but to keep simple we'll just despawn or hit immediately if close.
            if (p.target.hp <= 0) {
                this.projectiles.splice(i, 1);
                continue;
            }

            const dir = p.target.pos.sub(p.pos);
            const dist = dir.mag();
            const moveDist = p.speed * dt;

            if (dist <= moveDist) {
                // Hit!
                p.target.takeDamage(p.damage);
                if (p.onHit) p.onHit();
                this.projectiles.splice(i, 1);
            } else {
                p.pos = p.pos.add(dir.normalize().mul(moveDist));
            }
        }

        // Remove dead entities and spawn death effects
        for (let i = this.entities.length - 1; i >= 0; i--) {
            const e = this.entities[i];
            if (e.hp <= 0) {
                if (e.stats.name === 'Knight' || e.stats.name === 'Elite Knight') {
                    this.spawnEffect('particle_sword', e.pos, 1.4);
                } else if (e.stats.name !== 'Princess Tower' && e.stats.name !== 'King Tower') {
                    this.spawnEffect('death_ground_elixir1', e.pos, 0.7);
                }
                this.entities.splice(i, 1);
            }
        }

        // Update effects
        this.effects = this.effects.filter(e => this.timeElapsed - e.startTime < e.duration);

        this.resolveCollisions();
        this.constrainEntities();
    }

    spawnEffect(name: string, pos: Vector2, duration: number = 1.0) {
        console.log('SPAWNING EFFECT', name, pos);
        this.effects.push({
            id: this.nextEffectId++,
            name,
            pos: pos.clone(),
            startTime: this.timeElapsed,
            duration
        });
    }

    rivers = [
        { x: -5, y: CONFIG.RIVER_Y_START, width: CONFIG.LEFT_BRIDGE_X + 5, height: CONFIG.RIVER_Y_END - CONFIG.RIVER_Y_START },
        { x: CONFIG.LEFT_BRIDGE_X + CONFIG.BRIDGE_WIDTH, y: CONFIG.RIVER_Y_START, width: CONFIG.RIGHT_BRIDGE_X - (CONFIG.LEFT_BRIDGE_X + CONFIG.BRIDGE_WIDTH), height: CONFIG.RIVER_Y_END - CONFIG.RIVER_Y_START },
        { x: CONFIG.RIGHT_BRIDGE_X + CONFIG.BRIDGE_WIDTH, y: CONFIG.RIVER_Y_START, width: CONFIG.ARENA_WIDTH + 5 - (CONFIG.RIGHT_BRIDGE_X + CONFIG.BRIDGE_WIDTH), height: CONFIG.RIVER_Y_END - CONFIG.RIVER_Y_START }
    ];

    constrainEntities() {
        for (const entity of this.entities) {
            // Keep in bounds
            entity.pos.x = Math.max(entity.stats.radius, Math.min(CONFIG.ARENA_WIDTH - entity.stats.radius, entity.pos.x));
            entity.pos.y = Math.max(entity.stats.radius, Math.min(CONFIG.ARENA_HEIGHT - entity.stats.radius, entity.pos.y));

            // River collision via AABB
            const isDashing = entity.currentAbilityEffect === 'dashing_dash';
            if (!entity.stats.isAir && !entity.stats.jumpsRiver && !isDashing) {
                for (const river of this.rivers) {
                    const closestX = Math.max(river.x, Math.min(entity.pos.x, river.x + river.width));
                    const closestY = Math.max(river.y, Math.min(entity.pos.y, river.y + river.height));

                    const dx = entity.pos.x - closestX;
                    const dy = entity.pos.y - closestY;
                    const distSq = dx * dx + dy * dy;
                    
                    if (distSq < entity.stats.radius * entity.stats.radius) {
                        if (distSq === 0) {
                            // Center is exactly inside, push out to nearest edge
                            const distToLeft = entity.pos.x - river.x;
                            const distToRight = (river.x + river.width) - entity.pos.x;
                            const distToTop = entity.pos.y - river.y;
                            const distToBottom = (river.y + river.height) - entity.pos.y;
                            
                            const min = Math.min(distToLeft, distToRight, distToTop, distToBottom);
                            if (min === distToLeft) entity.pos.x = river.x - entity.stats.radius;
                            else if (min === distToRight) entity.pos.x = river.x + river.width + entity.stats.radius;
                            else if (min === distToTop) entity.pos.y = river.y - entity.stats.radius;
                            else entity.pos.y = river.y + river.height + entity.stats.radius;
                        } else {
                            const dist = Math.sqrt(distSq);
                            const overlap = entity.stats.radius - dist;
                            entity.pos.x += (dx / dist) * overlap;
                            entity.pos.y += (dy / dist) * overlap;
                        }
                    }
                }
            }
        }
    }

    addProjectile(sourceId: string, pos: Vector2, target: Entity, damage: number, speed: number, team: 'blue' | 'red', onHit?: () => void) {
        this.projectiles.push({
            id: this.nextProjectileId++,
            sourceId,
            startPos: pos.clone(),
            pos: pos.clone(),
            target,
            damage,
            speed,
            team,
            onHit
        });
    }

    resolveCollisions() {
        // Advanced steering behaviors based on official CSV Mass and CollisionRadius
        for (let i = 0; i < this.entities.length; i++) {
            for (let j = i + 1; j < this.entities.length; j++) {
                const a = this.entities[i];
                const b = this.entities[j];

                // Skip if both are buildings/towers or both are air
                if (a.stats.speed === 0 && b.stats.speed === 0) continue;
                if (a.stats.isAir && b.stats.isAir) continue; // Air units can overlap
                if (a.stats.isAir !== b.stats.isAir) continue; // Air and ground don't collide

                const distSq = a.pos.distanceSquaredTo(b.pos);
                // Convert integer CSV collision radius to tiles (e.g. 750 -> 0.75, 500 -> 0.50)
                const radiusA = a.stats.radius; 
                const radiusB = b.stats.radius;
                const minDist = radiusA + radiusB;
                
                if (distSq < minDist * minDist && distSq > 0.0001) {
                    if (a.stats.mass === Infinity && b.stats.mass === Infinity) {
                        continue; // Both static, don't move
                    }
                    
                    const dist = Math.sqrt(distSq);
                    const overlap = minDist - dist;
                    const dir = a.pos.sub(b.pos).normalize();

                    let ratioA = 0.5, ratioB = 0.5;
                    if (a.stats.mass === Infinity) {
                        ratioA = 0; ratioB = 1;
                    } else if (b.stats.mass === Infinity) {
                        ratioA = 1; ratioB = 0;
                    } else {
                        // Official Mass logic: Heavy units completely push aside light units.
                        // Mass scale: Giant = 18, Knight = 6, Skeleton = 1
                        const totalMass = a.stats.mass + b.stats.mass;
                        ratioA = b.stats.mass / totalMass;
                        ratioB = a.stats.mass / totalMass;
                        
                        // Amplify the mass difference to mimic the strict engine repulsion
                        if (a.stats.mass >= b.stats.mass * 3) {
                            ratioA = 0.05; ratioB = 0.95; // A dominates B
                        } else if (b.stats.mass >= a.stats.mass * 3) {
                            ratioA = 0.95; ratioB = 0.05; // B dominates A
                        }
                    }

                    if (ratioA > 0) a.pos = a.pos.add(dir.mul(overlap * ratioA));
                    if (ratioB > 0) b.pos = b.pos.sub(dir.mul(overlap * ratioB));
                    
                    // Sliding along buildings
                    if (a.stats.mass === Infinity || b.stats.mass === Infinity) {
                        const troop = a.stats.mass === Infinity ? b : a;
                        const building = a.stats.mass === Infinity ? a : b;
                        let pushDir = troop.pos.sub(building.pos).normalize();
                        if (Math.abs(pushDir.x) < 0.1) {
                            pushDir.x = pushDir.x >= 0 ? 0.3 : -0.3;
                            troop.pos = troop.pos.add(pushDir.normalize().mul(overlap));
                        }
                    }

                    // Flag for retargeting if physically displaced significantly
                    if (overlap * ratioA > 0.1) a.wasNudged = true;
                    if (overlap * ratioB > 0.1) b.wasNudged = true;
                }
            }
        }
    }
}

File: openroyale_sim/game.py
from __future__ import annotations

from dataclasses import dataclass
import math
import random
import string
from typing import Callable, Literal

from . import config
from .cards import CARDS, EntityStats, Team
from .entity import Entity
from .vector import Vector2


@dataclass
class Projectile:
    id: int
    source_id: str
    start_pos: Vector2
    pos: Vector2
    target: Entity
    damage: float
    speed: float
    team: Team
    splash_radius: float | None = None
    asset: str | None = None
    trajectory: Literal["line", "parabola"] | None = None
    on_hit: Callable[[], None] | None = None


@dataclass
class VisualEffect:
    id: int
    name: str
    pos: Vector2
    start_time: float
    duration: float
    file_name: str | None = None


@dataclass
class AoeCircle:
    id: int
    pos: Vector2
    radius: float
    start_time: float
    duration: float


class Game:
    def __init__(self, seed: int | None = None, cards: dict[str, EntityStats] | None = None) -> None:
        self.rng = random.Random(seed)
        self.cards = cards or CARDS
        self.entities: list[Entity] = []
        self.projectiles: list[Projectile] = []
        self.effects: list[VisualEffect] = []
        self.aoe_circles: list[AoeCircle] = []
        self.next_entity_id = 1
        self.next_projectile_id = 1
        self.next_effect_id = 1
        self.time_elapsed = 0.0
        self.rivers = [
            {"x": -5.0, "y": config.RIVER_Y_START, "width": config.LEFT_BRIDGE_X + 5.0, "height": config.RIVER_Y_END - config.RIVER_Y_START},
            {
                "x": config.LEFT_BRIDGE_X + config.BRIDGE_WIDTH,
                "y": config.RIVER_Y_START,
                "width": config.RIGHT_BRIDGE_X - (config.LEFT_BRIDGE_X + config.BRIDGE_WIDTH),
                "height": config.RIVER_Y_END - config.RIVER_Y_START,
            },
            {
                "x": config.RIGHT_BRIDGE_X + config.BRIDGE_WIDTH,
                "y": config.RIVER_Y_START,
                "width": config.ARENA_WIDTH + 5.0 - (config.RIGHT_BRIDGE_X + config.BRIDGE_WIDTH),
                "height": config.RIVER_Y_END - config.RIVER_Y_START,
            },
        ]

    def random_group_id(self) -> str:
        return "".join(self.rng.choice(string.ascii_lowercase + string.digits) for _ in range(7))

    def start(self) -> None:
        self.add_entity(self.cards["king_tower"], "blue", Vector2(9, 29))
        self.add_entity(self.cards["princess_tower"], "blue", Vector2(3, 25))
        self.add_entity(self.cards["princess_tower"], "blue", Vector2(15, 25))
        self.add_entity(self.cards["king_tower"], "red", Vector2(9, 3))
        self.add_entity(self.cards["princess_tower"], "red", Vector2(3, 7))
        self.add_entity(self.cards["princess_tower"], "red", Vector2(15, 7))

    def add_entity(self, stats: EntityStats, team: Team, pos: Vector2) -> Entity:
        if stats.spawn_count and stats.spawn_count > 1:
            first_entity: Entity | None = None
            radius = stats.spawn_radius or 1.0
            swarm_group_id = self.random_group_id()
            for i in range(stats.spawn_count):
                angle = (i / stats.spawn_count) * math.pi * 2
                offset = Vector2(math.cos(angle) * radius, math.sin(angle) * radius)
                entity = Entity(self.next_entity_id, stats, team, pos.add(offset), self)
                self.next_entity_id += 1
                entity.clone_group_id = swarm_group_id
                self.entities.append(entity)
                if first_entity is None:
                    first_entity = entity
            assert first_entity is not None
            return first_entity

        entity = Entity(self.next_entity_id, stats, team, pos, self)
        self.next_entity_id += 1
        self.entities.append(entity)
        return entity

    def add_entity_by_id(self, card_id: str, team: Team, pos: Vector2) -> Entity | None:
        stats = self.cards.get(card_id)
        if not stats:
            return None
        return self.add_entity(stats, team, pos)

    def update(self, dt: float) -> None:
        self.time_elapsed += dt

        for entity in list(self.entities):
            entity.update(dt)

        for i in range(len(self.projectiles) - 1, -1, -1):
            projectile = self.projectiles[i]
            if projectile.target.hp <= 0:
                self.projectiles.pop(i)
                continue

            direction = projectile.target.pos.sub(projectile.pos)
            dist = direction.mag()
            move_dist = projectile.speed * dt
            if dist <= move_dist:
                if projectile.splash_radius:
                    for entity in self.entities:
                        if entity.team != projectile.team and entity.hp > 0:
                            if entity.pos.distance_to(projectile.pos) <= projectile.splash_radius + entity.stats.radius:
                                entity.take_damage(projectile.damage)
                    self.aoe_circles.append(AoeCircle(self.next_effect_id, projectile.pos.clone(), projectile.splash_radius, self.time_elapsed, 0.3))
                    self.next_effect_id += 1
                else:
                    projectile.target.take_damage(projectile.damage)
                if projectile.on_hit:
                    projectile.on_hit()
                self.projectiles.pop(i)
            else:
                self.projectiles[i].pos = projectile.pos.add(direction.normalize().mul(move_dist))

        for i in range(len(self.entities) - 1, -1, -1):
            entity = self.entities[i]
            if entity.hp <= 0:
                if entity.stats.name in ("Knight", "Elite Knight"):
                    self.spawn_effect("particle_sword", entity.pos, 1.4)
                elif entity.stats.name == "Princess Tower":
                    self.spawn_effect("Tower_destroyed_ground1", entity.pos, math.inf, "building_tower")
                elif entity.stats.name == "King Tower":
                    self.spawn_effect("Tower_destroyed_ground2", entity.pos, math.inf, "building_tower")
                else:
                    self.spawn_effect("death_ground_elixir1", entity.pos, 0.7)
                self.entities.pop(i)

        self.effects = [effect for effect in self.effects if self.time_elapsed - effect.start_time < effect.duration]
        self.aoe_circles = [circle for circle in self.aoe_circles if self.time_elapsed - circle.start_time < circle.duration]

        self.resolve_collisions()
        self.constrain_entities()

    def spawn_effect(self, name: str, pos: Vector2, duration: float = 1.0, file_name: str | None = None) -> None:
        self.effects.append(VisualEffect(self.next_effect_id, name, pos.clone(), self.time_elapsed, duration, file_name))
        self.next_effect_id += 1

    def add_projectile(
        self,
        source_id: str,
        pos: Vector2,
        target: Entity,
        damage: float,
        speed: float,
        team: Team,
        splash_radius: float | None = None,
        asset: str | None = None,
        trajectory: Literal["line", "parabola"] | None = None,
        on_hit: Callable[[], None] | None = None,
    ) -> None:
        self.projectiles.append(
            Projectile(
                self.next_projectile_id,
                source_id,
                pos.clone(),
                pos.clone(),
                target,
                damage,
                speed,
                team,
                splash_radius,
                asset,
                trajectory,
                on_hit,
            )
        )
        self.next_projectile_id += 1

    def constrain_entities(self) -> None:
        for entity in self.entities:
            entity.pos.x = max(entity.stats.radius, min(config.ARENA_WIDTH - entity.stats.radius, entity.pos.x))
            entity.pos.y = max(entity.stats.radius, min(config.ARENA_HEIGHT - entity.stats.radius, entity.pos.y))

            is_dashing = entity.current_ability_effect == "dashing_dash"
            if not entity.stats.is_air and not entity.stats.jumps_river and not is_dashing:
                for river in self.rivers:
                    closest_x = max(river["x"], min(entity.pos.x, river["x"] + river["width"]))
                    closest_y = max(river["y"], min(entity.pos.y, river["y"] + river["height"]))
                    dx = entity.pos.x - closest_x
                    dy = entity.pos.y - closest_y
                    dist_sq = dx * dx + dy * dy
                    if dist_sq < entity.stats.radius * entity.stats.radius:
                        if dist_sq == 0:
                            dist_to_left = entity.pos.x - river["x"]
                            dist_to_right = river["x"] + river["width"] - entity.pos.x
                            dist_to_top = entity.pos.y - river["y"]
                            dist_to_bottom = river["y"] + river["height"] - entity.pos.y
                            minimum = min(dist_to_left, dist_to_right, dist_to_top, dist_to_bottom)
                            if minimum == dist_to_left:
                                entity.pos.x = river["x"] - entity.stats.radius
                            elif minimum == dist_to_right:
                                entity.pos.x = river["x"] + river["width"] + entity.stats.radius
                            elif minimum == dist_to_top:
                                entity.pos.y = river["y"] - entity.stats.radius
                            else:
                                entity.pos.y = river["y"] + river["height"] + entity.stats.radius
                        else:
                            dist = math.sqrt(dist_sq)
                            overlap = entity.stats.radius - dist
                            entity.pos.x += (dx / dist) * overlap
                            entity.pos.y += (dy / dist) * overlap

    def resolve_collisions(self) -> None:
        for i in range(len(self.entities)):
            for j in range(i + 1, len(self.entities)):
                a = self.entities[i]
                b = self.entities[j]
                if a.stats.speed == 0 and b.stats.speed == 0:
                    continue
                if a.stats.is_air and b.stats.is_air:
                    continue
                if a.stats.is_air != b.stats.is_air:
                    continue

                dist_sq = a.pos.distance_squared_to(b.pos)
                min_dist = a.stats.radius + b.stats.radius
                if dist_sq < min_dist * min_dist and dist_sq > 0.0001:
                    if math.isinf(a.stats.mass) and math.isinf(b.stats.mass):
                        continue

                    dist = math.sqrt(dist_sq)
                    overlap = min_dist - dist
                    direction = a.pos.sub(b.pos).normalize()

                    ratio_a = 0.5
                    ratio_b = 0.5
                    if math.isinf(a.stats.mass):
                        ratio_a = 0
                        ratio_b = 1
                    elif math.isinf(b.stats.mass):
                        ratio_a = 1
                        ratio_b = 0
                    else:
                        total_mass = a.stats.mass + b.stats.mass
                        ratio_a = b.stats.mass / total_mass
                        ratio_b = a.stats.mass / total_mass
                        if a.stats.mass >= b.stats.mass * 3:
                            ratio_a = 0.05
                            ratio_b = 0.95
                        elif b.stats.mass >= a.stats.mass * 3:
                            ratio_a = 0.95
                            ratio_b = 0.05

                    if ratio_a > 0:
                        a.pos = a.pos.add(direction.mul(overlap * ratio_a))
                    if ratio_b > 0:
                        b.pos = b.pos.sub(direction.mul(overlap * ratio_b))

                    if math.isinf(a.stats.mass) or math.isinf(b.stats.mass):
                        troop = b if math.isinf(a.stats.mass) else a
                        building = a if math.isinf(a.stats.mass) else b
                        push_dir = troop.pos.sub(building.pos).normalize()
                        if abs(push_dir.x) < 0.1:
                            push_dir.x = 0.3 if push_dir.x >= 0 else -0.3
                            troop.pos = troop.pos.add(push_dir.normalize().mul(overlap))

                    if overlap * ratio_a > 0.1:
                        a.was_nudged = True
                    if overlap * ratio_b > 0.1:
                        b.was_nudged = True
